internal/util/image/disk_cache.go

```diff
@@ -0,0 +1,144 @@
+package image
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"iter"
+	"os"
+	"path/filepath"
+	"slices"
+	"time"
+
+	"github.com/containerd/containerd/archive"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/opencontainers/go-digest"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	errorutil "github.com/operator-framework/operator-controller/internal/util/error"
+	fsutil "github.com/operator-framework/operator-controller/internal/util/fs"
+)
+
+const ConfigDirLabel = "operators.operatorframework.io.index.configs.v1"
+
+func CatalogCache(basePath string) Cache {
+	return &diskCache{
+		basePath: basePath,
+		filterFunc: func(ctx context.Context, srcRef reference.Named, image ocispecv1.Image) (archive.Filter, error) {
+			_, specIsCanonical := srcRef.(reference.Canonical)
+			dirToUnpack, ok := image.Config.Labels[ConfigDirLabel]
+			if !ok {
+				// If the spec is a tagged ref, retries could end up resolving a new digest, where the label
+				// might show up. If the spec is canonical, no amount of retries will make the label appear.
+				// Therefore, we treat the error as terminal if the reference from the spec is canonical.
+				return nil, errorutil.WrapTerminal(fmt.Errorf("catalog image is missing the required label %q", ConfigDirLabel), specIsCanonical)
+			}
+
+			return allFilters(
+				onlyPath(dirToUnpack),
+				forceOwnershipRWX(),
+			), nil
+		},
+	}
+}
+
+func BundleCache(basePath string) Cache {
+	return &diskCache{
+		basePath: basePath,
+		filterFunc: func(_ context.Context, _ reference.Named, _ ocispecv1.Image) (archive.Filter, error) {
+			return forceOwnershipRWX(), nil
+		},
+	}
+}
```

Comment:
These filter funcs are literally the only difference between catalog and bundle image unpacking. :)